### PR TITLE
File logger and file log sharing added

### PR DESCRIPTION
**Pull request by pawelgil**
# What was done?

Logs are being stored to a file shareable via **debug** or **beta** settings menu.

IMG_0717 (1)

IMG_0718

example log file:

[log.txt](https://github.com/HabitatMap/AirCastingiOS/files/8010467/log.txt)

As we move forward and the app gets more complex we are going to need some more advanced remote debugging tools. This is a first step in this process.

# Technical notes

* The file will be trimmed from the beginning periodically so it doesn't grow indefinitely
* File writes are buffered (current buffer size is 25 and can be adjusted)
* OOP design to ensure testability and flexibility

# Object graph

Not full. Just a rough map to grasp the idea

image

# GarbageCollector

Also introduced in this PR, `GarbageCollector` is an object that frees disk space upon receiving system notification asking the app to.

### CODE REVIEW

File: AirCasting/Utils/Logging/FileLogger/FileLoggerHeaderProvider.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                                  P@@@@&5  5@@@@@5
+                                 5&@@@@P    G@@@@&
+                                 #@@@@B      #@@@@B
+                                B@@@@@&&&&&&&&@@@@@G
+                               P@@@@@@@@@@@@@@@@@@@@5
+                              5&@@@@G5555555555B@@@@&
+                              #@@@@#            &@@@@B
+                             B@@@@&5            P@@@@@G
+                             PPPPP5              PPPPPP
+"""
```
Comment (jankrzempek) [inline AirCasting/Utils/Logging/FileLogger/FileLoggerHeaderProvider.swift:40]: 🚀

File: AirCasting/Utils/Logging/LoggerBuilder.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @discardableResult
+    func dispatchOn(_ queue: DispatchQueue) -> Self {
+        partialLogger = ScheduledLogger(queue: queue, logger: partialLogger)
+        return self
+    }
+    
+    func build() -> Logger {
+        return partialLogger
+    }
+
```
Comment (jankrzempek) [inline AirCasting/Utils/Logging/LoggerBuilder.swift:38]: You can remove one line I think, if not please omit 😄

**Review by jankrzempek — APPROVED** [review]
🚀✨

File: AirCasting/Settings/SettingsView.swift [hunk 3]
```diff
@@ -252,5 +270,11 @@ struct SettingsView: View {
         })
             .font(Fonts.boldHeading1)
     }
+    
+    private var shareLogsButton: some View {
+        Button("Share logs") {
```
[the remark this answers is not in the document]
Comment (pawelgil) [reply AirCasting/Settings/SettingsView.swift:275]: You are correct - BETA/DEBUG only strings should not go to `Strings` as we use them internally and we don't want to waste effort on translating those.

**Pull request merged**